### ISSUE TITLE
Build failing

### ISSUE THREAD

**Issue by AshotN** [socketio/socket.io-client-cpp#212]
I'm getting a ton of errors saying

> overriding virtual function has less restrictive exception specification than base class virtual member function

Using Visual Studio 17 on Windows 10

**Comment by Jackey-Huo** [comment]
@AshotN I got the same error. Have you found any solution about it?

**Comment by AshotN** [comment]
I gave up on the project I was working on related to this. Sorry I couldn't be more helpful

**Comment by Jackey-Huo** [comment]
I've resolved the bug. This repo depend on [websocketpp](https://github.com/zaphoyd/websocketpp) as submodules. The websocktpp version has been out of date. update the websocktpp to latest commit will resolve all trouble ;)

**Comment by lastpeony** [comment]
> I've resolved the bug. This repo depends on [websocketpp](https://github.com/zaphoyd/websocketpp) as submodules. The websocktpp version has been out of date. update the websocktpp to latest commit will resolve all trouble ;)

Hello i am facing with the same problem.How do i  exactly update websocktpp?

**Comment by Jackey-Huo** [comment]
using the following command:

```
    cd lib/websocketpp
    git pull origin master
```

Then rebuild the whole project.

Good Luck ;)

**Comment by keniiyoshi** [comment]
I did this for building the INSTALL project, and it worked!
However, I got these set of new errors, which repeated 3 times in sioclient project, io_object_impl.hpp line 87:

C2039  'get_executor': is not a member of 'boost::reference_wrapper<boost::asio::io_service>'
C2665  'boost::asio::detail::io_object_executor<Executor>::io_object_executor': none of the 3 overloads could convert all the argument types
        with
        [
            Executor=boost::asio::executor
        ]	

Did you come across these errors and found a fix?

I'm using Visual Studio 19 on Windows 10.

**Comment by Jackey-Huo** [comment]
hi Ken,

unfortunately, I've never come across this error. but regard to the error
message you provided, it might be caused by unmatch version or installation
of boost lib.

bests and good luck ;)

Ken Iiyoshi <notifications@github.com> 于 2019年7月9日周二 下午8:17写道：

**Comment by coreyp1** [comment]
I confirm the new error that keniiyoshi is reporting.  I updated my websocketpp according to Jackey-Huo's instructions, and, once I `#include<sio_client.h>` (as in, just that single line will trigger the problem), the new errors show up.  I'm using VS2019.

**Comment by coreyp1** [comment]
Whew.  Finally figured out what was happening.

1. websocketpp, even after being updated, absolutely **will not** compile with anything after **boost 1.69**.  (I was originally trying to use Boost 1.71.

2. **Boost 1.69** pre-built binaries for Windows don't exist for **msvc-14.2** (e.g., VS2019), and I couldn't get the **msvc-14.1** ones (originally for VS2017) to work with VS2019. Maybe it's a PEBCAK issue, but I'm new to Visual Studio in general.

3. Boost 1.69 build tools doesn't support VS2019.

4. Basically, it took a 3 days of messing around to finally get Boost 1.69 to compile under VS2019 (msvc-14.2).

After all of that, socket.io wouldn't compile, because in 2 places in `sio_client_impl.cpp`, I kept getting compiler error **C3321 "An initializer list is unexpected in this context"**.

Basically, there were two lines like this: `LOG("blah blah blah"){};`

The compiler doesn't like it.  I had to delete the spurious `{}`, and then everything finally compiled.

I'm leaving this comment here in case it helps someone else.

**Comment by Jackey-Huo** [comment]
@coreyp1 Congratulation! It seems you have experienced a quite struggle days ;)
BTW is this Issue could be regarded as solved and shall we close it? @AshotN @melode11

**Comment by coreyp1** [comment]
I don't think that it was "solved", but rather "worked around".  I see 3 issues, in which 1 can't be fixed, and 2 which should be patched.

1. Websocketpp doesn't work with Boost after 1.69.  See https://github.com/zaphoyd/websocketpp/issues/808 and https://github.com/zaphoyd/websocketpp/issues/794.  Obviously, this project can't do anything about that, and must wait for websocketpp to fix their issues.

2. The version of Websocketpp library that is linked to is from 2015.  This should at least be updated in order to eliminate the compile error message originally reported at the top of this issue.  At the very least, even if the update doesn't solve the Boost version problem, it does eliminate a huge group of compile errors in VS2017 and VS2019.

3. In sio_client_impl.cpp (https://github.com/socketio/socket.io-client-cpp/blob/master/src/internal/sio_client_impl.cpp), the extra `{}` should be removed, as it is incompatible syntax for modern compilers.  Specifically, line 282 should change from
`LOG("ping exit,con is expired?"<<m_con.expired()<<",ec:"<<ec.message()<<endl){};`
to
`LOG("ping exit,con is expired?"<<m_con.expired()<<",ec:"<<ec.message()<<endl);`
And line 496 should change from
`if(ec)LOG("ec:"<<ec.message()<<endl){};`
to
`if(ec)LOG("ec:"<<ec.message()<<endl);`
or
`if(ec){LOG("ec:"<<ec.message()<<endl)};`
depending on your coding style.

I don't think that this issue should be marked as resolved until 2 and 3 are done, though, because that is what it will take to make the project compile.  The Boost 1.69 situation is merely a linking issue.

**Comment by rthickling** [comment]
Note that I see [zaphoyd](https://github.com/zaphoyd) has [pointed out](https://github.com/zaphoyd/websocketpp/issues/808#issuecomment-616200539) that in [websocketpp](https://github.com/zaphoyd/websocketpp) "The fix for this issue has been available on the develop branch for some time and has now been published in release 0.8.2".

**Comment by richardeakin** [comment]
Just want to point out that 2.x branch still needs an update for the `LOG()..{}` compile error in sio_client_impl.cpp. Thanks for documenting that fix @coreyp1

**Comment by darrachequesne** [comment]
This should be fixed now. Please reopen if needed.
